drivers/pn532/pn532.c

```diff
@@ -0,0 +1,639 @@
+
+/* Mifare specific commands */
+#define MIFARE_CMD_READ               (0x30)
+#define MIFARE_CMD_WRITE              (0xA0)
+
+/* RF register settings */
+#define RF_CONFIG_MAX_RETRIES         (0x05)
+
+/* Buffer operations */
+#define BUFF_CMD_START                (6)
+#define BUFF_DATA_START               (BUFF_CMD_START + 1)
+#define RAPDU_DATA_BEGIN              (1)
+#define RAPDU_MAX_DATA_LEN            (PN532_BUFFER_LEN - BUFF_DATA_START - 5)
+#define CAPDU_MAX_DATA_LEN            (PN532_BUFFER_LEN - BUFF_DATA_START - 1)
+
+/* Length for passive listings */
+#define LIST_PASSIVE_LEN_14443(num)   (num * 20)
+
+#if ENABLE_DEBUG
+#define PRINTBUFF printbuff
+static void printbuff(char *buff, unsigned len)
+{
+    while (len) {
+        len--;
+        printf("%02x ", *buff++);
+    }
+    printf("\n");
+}
+#else
+#define PRINTBUFF(...)
+#endif
+
+static void _nfc_event(void *dev)
+{
+    mutex_unlock(&((pn532_t *)dev)->trap);
+}
+
+void pn532_reset(pn532_t *dev)
+{
+    DEBUG("pn532: reset\n");
+    gpio_clear(dev->conf->reset);
+    xtimer_usleep(400000);
+    gpio_set(dev->conf->reset);
+    xtimer_usleep(10000);
+}
+
+int pn532_init(pn532_t *dev, const pn532_params_t *params)
+{
+    int ret;
+
+    dev->conf = params;
+
+    gpio_init_int(dev->conf->irq, GPIO_IN_PU, GPIO_FALLING,
+                  _nfc_event, (void *)dev);
+
+    gpio_init(dev->conf->reset, GPIO_OUT);
+    gpio_set(dev->conf->reset);
+
+    ret = i2c_init_master(dev->conf->i2c, I2C_SPEED_NORMAL);
+    if (ret == 0) {
+        pn532_reset(dev);
+    }
+
+    mutex_init(&dev->trap);
+    mutex_lock(&dev->trap);
+
+    return ret;
+}
+
+static unsigned char chksum(char *b, unsigned len)
+{
+    unsigned char c = 0x00;
+
+    while (len--) {
+        c -= *b++;
+    }
+    return c;
+}
+
+static int _write(pn532_t *dev, char *buff, unsigned len)
+{
+    int ret;
+
+    i2c_acquire(dev->conf->i2c);
+    DEBUG("pn532: -> ");
+    PRINTBUFF(buff, len);
+    ret = i2c_write_bytes(dev->conf->i2c, PN532_I2C_ADDRESS, buff, len);
+    i2c_release(dev->conf->i2c);
+    return ret;
+}
+
+static int _read(pn532_t *dev, char *buff, unsigned len)
+{
+    /* len, +1 for 0x01 after read is accepted */
+    int ret;
+
+    i2c_acquire(dev->conf->i2c);
+    ret = i2c_read_bytes(dev->conf->i2c, PN532_I2C_ADDRESS, buff, len + 1);
+    i2c_release(dev->conf->i2c);
+
+    DEBUG("pn532: <- ");
+    PRINTBUFF(buff, r);
+    return ret;
+}
+
+static int send_cmd(pn532_t *dev, char *buff, unsigned len)
+{
+/*
+ * PN532 information frame format
+ *
+ * 0  1  2  3  4  5  6            M
+ * +--+--+--+--+--+--+--+--+---+--+--+--+
+ * |00|00|FF|LP|LC|FI|D0|D1|...|Dn|CS|00|
+ * +--+--+--+--+--+--+--+--+---+--+--+--+
+ * 00: 0x00
+ * FF: 0xFF
+ * LP: packet length (including FI)
+ * LC: packet length checksum [LEN + LCS] = 0x00
+ * FI: frame identifier (D4: host -> controller, D5: controller -> host)
+ * Dx: data
+ * CS: checksum [FI+D0+D1+...+Dn+CS] = 0x00
+ */
```

Comment:
Nope, this is the frame structure we need to send, not read. I could just remove the docu if it's misleading. :+1: